Fastsync English spec for review (#208)
informalsystems/tendermint-rs: docs/spec/fastsync/fastsync.md

# Fastsync

> <!--

Reviewer: Maybe we should remove these comments from the template.

Author: I removed them in a previous revision

  

# Part I - Outside view

## Context of this document

<!--
> mention other components and or specifications that are relevant for this
spec. Possible interactions, possible use cases, etc. 
---->
<!--
> should give the reader the understanding in what environment this component
will be used. 
---->

Fastsync is a protocol that is used by a full node to catch-up to the
current state of a Tendermint blockchain. Its typical use case is a
full node that was disconnected from the system for some time. The
recovering full node locally has a copy of a prefix of the blockchain,
and the corresponding application state that is slightly outdated. It
then queries its peers for the blocks that were decided on by the
Tendermint blockchain during the period the full node was
disconnected. After receiving these blocks, it executes the
transactions in the blocks in order to catch-up to the current height
of the blockchain and the corresponding application state.

In practice it is sufficient to catch-up only close to the current
height: The Tendermint consensus reactor implements a similar

> functionality and can synchronize a node that is approximately 100

Reviewer: I don't remember seeing code that does this in consensus reactor. @ebuchman @melekes Can someone confirm this? Ideally we should also add a reference to the corresponding code here.

Reviewer 2: Consensus reactor can catch up, but it's very slow comparing to fast sync & state sync.

Author: Ethan changed it to 10 already.

the relevant fields of the block, for the purpose of this
specification, we will assume that the blockchain is a list of
headers, rather than a list of blocks. 

#### **[TMBC-SEQ]**:

The Tendermint blockchain is a list *chain* of headers. 

#### **[TMBC-SEQ-GROW]**: 

During operation, new headers may be appended to the list one by one.

> In the following, *ETIME* is a lower bound
> on the time interval between the times at which two
> successor blocks are added. 

#### **[TMBC-SEQ-APPEND-E]**: 
If a header is appended at time *t* then no additional header will be
appended before time *t + ETIME*.

#### **[TMBC-Auth-Byz]**:
The authenticated Byzantine model assumes that no node (faulty or
correct) may break digital signatures and in addition to this, no
assumption is made about the internal behavior of faulty full
nodes. That is, faulty nodes are limited in that they cannot forge
messages.

<!-- The authenticated Byzantine model assumes [TMBC-Sign-NoForge] and -->
<!-- [TMBC-FaultyFull], that is, faulty nodes are limited in that they -->
<!-- cannot forge messages [TMBC-Sign-NoForge]. -->

#### **[TMBC-VALIDATOR-Set]**:

> A *validator set* is a set of validator pairs. For a validator set

Author: The problem was, that in the existing documentation, "validator" refers to both a data structure and a full node. I added the definition. Validator pair refers to the data structure (Address + voting power) and validator node is the full node.

*vs*, we write *TotalVotingPower(vs)* for the sum of the voting powers
of its validator pairs.

#### **[TMBC-CORRECT]**:
We define a predicate *correctUntil(n, t)*, where *n* is a node and *t* is a 
time point. 
The predicate *correctUntil(n, t)* is true if and only if the node *n* 
follows all the protocols (at least) until time *t*.

#### **[TMBC-TIME-PARAMS]**:
A blockchain has the following configuration parameters:
 - *unbondingPeriod*: a time duration.
 - *trustingPeriod*: a time duration smaller than *unbondingPeriod*.

#### **[TMBC-FM-2THIRDS]**:
If a block *h* is in the chain,
then there exists a subset *CorrV*
of *h.NextValidators*, such that:
  - *TotalVotingPower(CorrV) > 2/3
    TotalVotingPower(h.NextValidators)*; 
  - For every validator pair *(n,p)* in *CorrV*, it holds *correctUntil(n,
    h.Time + trustingPeriod)*.

#### **[TMBC-CorrFull]**: 
Every correct full node locally stores a prefix of the

> current list of headers from [**[TMBC-SEQ]**](TMBC-SEQ-link).

Reviewer 3: ```suggestion
current list of headers from [**[TMBC-SEQ]**][TMBC-SEQ-link].
```
the link is broken

## Informal Problem statement

<!--
> for the general audience, that is, engineers who want to get an overview over what the component is doing
from a bird's eye view. 
---->
A full node has as input a block of the blockchain at height *h* and
the corresponding application state (or the prefix of the current
blockchain until height *h*). It has access to a set *peerIDs* of full
nodes called *peers* that it knows of.  The full node uses the peers
to read blocks of the Tendermint blockchain (in a safe way, that is,
it checks the soundness conditions), until it has read the most recent
block and then terminates.

## Sequential Problem statement

<!--
> should be English and precise. will be accompanied with a TLA spec.
---->

*Fastsync* gets as input a block of height *h* and the corresponding
application state *s* that corresponds to the block and state of that

> height of the blockchain [**[TMBC-SEQ]**][TMBC-SEQ-link], and produces

Author: I guess it is not. I removed the reference.

as output (i) a list *L* of blocks starting at height *h* to some height
*terminationHeight*, and (ii) the application state when applying the
transaction of the list *L* to *s*. Fastsync has to satisfy the following

> properties [FS-Seq-?]:

Author: I wanted to use '?' as wildcard but I guess it is not clear. Any idea how to write that?

> properties [FS-Seq-?]:

Reviewer 4: Once we have traceability identifiers around, patterns like `?` will be unparseable by the tool.

> properties [FS-Seq-?]:

Author: Zarko suggested to remove that tag

> properties [FS-Seq-?]:

Reviewer 5: Ah, I'm used to * as a wildcard, but in any case think it can just be removed here

#### **[FS-Seq-Live]**: 
*Fastsync* eventually terminates.

 
#### **[FS-Seq-Term]**:
Let *bh* be the height of the blockchain at the time *Fastsync* starts.
When *Fastsync* terminates, it outputs a list of all blocks from

> height *h* to some height *terminationHeight >= bh - 1*,

Author: Yes. Good point. I added a clarification right there.

> [**[TMBC-SEQ]**][TMBC-SEQ-link].

Author: I agree. It is not so clear (and not consistently throughout the document). I removed the references where they were unclear.

----

 ```go
func Block(addr Address, height int64) (Block, error)
```
- Implementation remark
   - RPC to full node *addr*
- Expected precondition
  - 'height` is less than or equal to height of the peer
- Expected postcondition
  - if *addr* is correct: Returns the block of height `height`
  from the blockchain. [FS-A-COMM]
  - if *addr* is faulty: Returns arbitrary block [**[TMBC-Auth-Byz]**][TMBC-Auth-Byz-link]
- Error condition
  - if *addr* is correct: precondition violated. [FS-A-COMM]
  - if *addr* is faulty: arbitrary error (including timeout). [**[TMBC-Auth-Byz]**][TMBC-Auth-Byz-link]
----

### Temporal Properties

#### Fairness

As mentioned above, without assumptions on the correctness of some
peers, no protocol can achieve the required specifications. Therefore, 
we sometimes consider the following (fairness) constraint in the
safety and liveness properties below:

#### **[FS-SOME-CORR-PEER]**:
Initially, the set *peerIDs* contains at least one correct full node.

> While in principle the above condition can be part of a sufficient
> condition to solve [FS-Seq-Live] and
> [FS-Seq-Term] and [FS-Seq-Term-SYNC], we will see below that the

> > current implementation of Fastsync (V2) requires the following (much

Author: As far as I understand the problem (or a similar one) also is present at the other implementations. We should discuss what precisely we can put here.

> > current implementation of Fastsync (V2) requires the following (much

Reviewer: It is also true for the past versions.

has not provided a block recently (check of *peerTimeStamp[p]*) or it
has not provided sufficiently many data (check of *peerRate[p]*), then
*p* is removed from *peerIDs*.
  
#### **[FS-V2-TIMEOUT]**:

*Fastsync V2* starts a timeout whenever a block is
executed (that is, when the height is incremented). If the timeout expires
before the next block is executed, *Fastsync* terminates.
We say that if *peerIDs* is empty upon termination, then *Fastsync* terminates
with failure, otherwise it terminates successfully.

### Details

<!--
> Function signatures followed by pseudocode (optional) and a list of features (required):
> - Implementation remarks (optional)
>   - e.g. (local/remote) function called in the body of this function
> - Expected precondition
> - Expected postcondition
> - Error condition
---->

```go
func QueryStatus()
```
- Expected precondition
    - peerIDs initialized and non-empty
- Expected postcondition
    - call asynchronously `Status(n)` at each peer *n* in *peerIDs*.
- Error condition
    - fails if precondition is violated
----

```go
func OnStatusResponse(addr Address, ht int64)
```
- Comment
    - *ht* is a height

> 	- peers can provide the status without being called

Author: I agree. I aimed at a more uniform terminology with e.g. the light client specs, and at some point I realized I needed that particular addition. The nice thing about functions is that I can describe expected pre- and postconditions very nicely. Let's discuss how to express what we need.

- Expected precondition
    - *peerHeights(addr) <= ht*
- Expected postcondition
    - *peerHeights(addr) = ht*
	- *TargetHeight* is updated
- Error condition

>     - if precondition is violated: *addr* not in *peerIDs* (that is,

Author: It means the latter, that is, "if it returns an error, then addr will be removed from peerIDs"

      *addr* is removed from *peerIDs*)
- Timeout condition
    - if `OnStatusResponse(addr, ht)` was not invoked within *2 Delta* after
	`Status(addr)` was called:  *addr* not in *peerIDs*
----

```go
func CreateRequest
```
- Expected precondition
    - *height < TargetHeight*
	- *peerIDs* nonempty
- Expected postcondition
    - Function `Block` is called remotely at a peer *addr* in peerIDs 
	  for a missing height *h*  
	  *Remark:* different implementations may have different
      strategies to balance the load over the peers
    - *pendingblocks(h) = addr*
----

```go
func OnBlockResponse(addr Address, b Block)
```
- Comment
    - if after adding block *b*, blocks of heights *height* and
      *height + 1* are in *blockstore*, then `Execute` is called
- Expected precondition
    - *pendingblocks(b.Height) = addr*
	- *b* satisfies basic soundness  
- Expected postcondition
    - if function `Execute` has been executed without error or was not executed:
        - *receivedBlocks(b.Height) = addr*

> 	    - *blockstore(b.Height) = b*

Author: Yes, this just describes bookkeeping that is in addition to what is done within Execute.

		- *peerTimeStamp[addr]* is set to a time between invocation and
          return of the function.
		- *peerRate[addr]* is updated according to size of received
          block
- Error condition
    - if precondition is violated: *addr* not in *peerIDs*; reset
	*pendingblocks(b.Height)* to nil;
- Timeout condition
    - if `OnBlockResponse(addr, b)` was not invoked within *2 Delta* after
	`Block(addr,h)` was called for *b.Height = h*: *addr* not in *peerIDs*
----

```go
func Execute()
```
- Comments
    - none
- Expected precondition
	- application state is the one of the blockchain at height
      *height - 1*

> 	- **[FS-V2-Verif]** for any two blocks *a* and *b* from

Author: No, I want this particular condition to be labeled for later reference.

####  **[FS-ISSUE-KILL]**:
If two blocks are not matching [FS-V2-Verif], `Execute` dismisses both
blocks and removes the peers that provided these blocks from
*peerIDs*. If block *a* was correct and provided by a correct peer *p*,
and block b was faulty and provided by a faulty peer, the protocol
- removes the correct peer *p*, although it might be useful to
  download blocks from it in the future
- removes the block *a*, so that a fresh copy of *a* needs to be downloaded
  again from another peer
  
By [FS-A-PEER] we do not put a restriction on the number
  of faulty peers, so that faulty peers can make *FS* to remove all
  correct peers from *peerIDs*. As a result, this version of
  *Fastsync* violates [FS-VC-CORR-INV-SYNC].

####  **[FS-ISSUE-NON-TERM]**:

Due to [**[FS-ISSUE-KILL]**](#fs-issue-kill), from some point on, only
faulty peers may be in *peerIDs*. They can thus control at which rate
*Fastsync* gets blocks. If the timeout duration from [FS-V2-TIMEOUT]
is greater than the time it takes to add a block to the blockchain
(LTIME in [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]), the
protocol may never terminate and thus violate [FS-VC-TERM].  This

> scenario is even possible if a correct peer is always in *peerIDs*,

Reviewer 5: Slash isn't it possible with just one correct peer if we manage to always eventually request pairs of sequential blocks from that correct peer?

> scenario is even possible if a correct peer is always in *peerIDs*,

Reviewer: I guess your comment applies to the case above: we could ask always consecutive blocks from a peer, but it would lead to waste of resources as you would need for every block to download two blocks (in case requests are made to different peers). More efficient solution would be asking block and commit from a single peer, but that would require Tendermint API.

from *peerIDs*.  That is, if
we sequentially verify blocks starting with *startBlock*, we will
never remove a correct peer from *peerIDs* and we will be able to
ensure the following invariant:

#### **[NewFS-VAR-PEER-INV]**:
If a peer never misbehaves, it is never removed from *peerIDs*. It
follows that under [FS-SOME-CORR-PEER], *peerIDs* is always non-empty.

> To perform these checks, we suggest to change the protocol as follows

#### Fastsync has the following configuration parameters:
- *trustingPeriod*: a time duration; cf.
  [**[TMBC-TIME-PARAMS]**][TMBC-TIME-PARAMS-link].

> [NewFS-A-INIT] is the suggested replacement of [FS-A-V2-INIT]. This will
> allow us to use the established trust to understand precisely which
> peer reported an invalid block in order to ensure the
> invariant [NewFS-VAR-TRUST-INV] below:

#### **[NewFS-A-INIT]**:
- *startBlock* is from the blockchain, and within *trustingPeriod*
(possible with some extra margin to ensure termination before
*trustingPeriod* expired)
- *startState* is the application state of the blockchain at Height
  *startBlock.Height*.
- *startHeight = startBlock.Height*

#### Additional Variables

> - *trustedBlockstore*: stores for each height greater than or equal to

Author: Do you mean to rename it?

    *startBlock.Height*, the block of that height. Initially it
    contains only *startBlock*

#### **[NewFS-VAR-TRUST-INV]**:
Let *b(i)* be the block in *trustedBlockstore*
with b(i).Height = i. It holds that
for *startHeight < i < height - 1*, 
*VerifyCommit (b(i),b(i+1).Commit) = true*.

> We propose to update the function `Execute`. To do so, we first
> define the following helper functions:

```go
func ValidCommit(VS ValidatorSet, C Commit) Boolean
```
- Comments
    - checks validator set based on [**[TMBC-FM-2THIRDS]**][TMBC-FM-2THIRDS-link]
- Expected precondition
	-  The validators in *C*
	     - are a subset of VS
		 - have more than 2/3 of the voting power in VS
- Expected postcondition
    - returns *true* if precondition holds, and *false* otherwise
- Error condition
    - none 
----

```go

> func SequentialVerify {

Author: We can discuss in what detail we want to describe possible fixes.

> func SequentialVerify {

Reviewer: In my view, it is sufficient to describe the fix at the high level, as done above, so we don't need to enter the details of the pseudocode.

	while (true) {
		b1 = blockstore[height];
		b2 = blockstore[height+1];
		if b1 == nil or b2 == nil {
			exit;
		}
		if ValidCommit(trustedBlockstore[height - 1].NextValidators, b2.commit) {

> 			// we trust b2

Reviewer: This follows from the definition above.

			if VerifyCommit(b1, b2.commit) {
				trustedBlockstore.Add(b1);
				height = height + 1;
			}
			else {
				// as we trust b2, b1 must be faulty
				blockstore.RemoveFromPeer(receivedBlocks[height]);
				// we remove all blocks received from the faulty peer
				peerIDs.Remove(receivedBlocks(bnew.Height));
				exit;
			
			}			
		} else {
			// b2 is faulty
			blockstore.RemoveFromPeer(receivedBlocks[height + 1]);
			// we remove all blocks received from the faulty peer
		    peerIDs.Remove(receivedBlocks(bnew.Height));
			exit;
			}
		}
}
```
- Comments
    - none
- Expected precondition
	- [NewFS-VAR-TRUST-INV]
- Expected postcondition
    - [NewFS-VAR-TRUST-INV]

> 	- there is no block *bnew* with *bnew.Height = height + 1* in

Author: This says that you don't stop as long as there is another block in blockstore, or in other words, you iterate to the maximal possible height at that moment.

some height *terminationHeight >= maxh*.

> We use *term - TD* as reference time, as we have to account
> for communication delay between the peer and *FS*. After the peer sent
> the last message to *FS*, the peer and *FS* run concurrently and
> independently. There is no assumption on the rate at which a peer can
> add blocks (e.g., it might be in the process of catching up itself).

>  (*TD* might depend on timeouts etc. We suggest that an acceptable
>  value for *TD* is in the range of apporx. 10 sec., that is the
>  intervall between two calls `QueryStatus()`; see below.

> Under [FS-ALL-CORR-PEER], if *peerIDs* contains a full node that is
> "synchronized with the blockchain", and *blockchainheight* is the height
> of the blockchain at time *term*, then  *terminationHeight* may even
> achieve
> *blockchainheight - TD / ETIME*;
> cf. [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]. 

#### Liveness

<!--
> liveness specifications in English. Possibly with timing/fairness requirements:
e.g., if the component is connected to a correct full node and communication is
reliable and timely, then something good happens eventually. 
---->

#### **[FS-VC-ALL-CORR-TERM]**:

> Under [FS-ALL-CORR-PEER], *Fastsync* eventually terminates successfully.

Reviewer: Where do we define what we mean by termination with failure?

Author: In the following. Is this OK?

#### **[FS-V2-TIMEOUT]**:

*Fastsync V2* starts a timeout whenever a block is
executed (that is, when the height is incremented). If the timeout expires
before the next block is executed, *Fastsync* terminates.
We say that if *peerIDs* is empty upon termination, then *Fastsync* terminates
with failure, otherwise it terminates successfully.

<!--
> In this section we become more concrete, with basic (abstracted) data types 
---->
<!--
> some math that allows to write specifications and pseudo code solution below.
Some variables, etc. 
---->

> We now introduce variables and auxiliary functions used by the protocol.

#### Inputs
- *startBlock*: the block Fastsync starts from
- *startState*: application state corresponding to *startBlock.Height*

#### **[FS-A-V2-INIT]**:
- *startBlock* is from the blockchain
- *startState* is the application state of the blockchain at Height *startBlock.Height*.

#### Variables
- *height*: initially *startBlock.Height + 1*
  > height should be thought of the "height of the next block we need to download"
- *state*: initially *startState*
- *peerIDs*: peer addresses [FS-A-PEER-IDS](#fs-a-peer-ids)
- *peerHeights*: stores for each peer the height it reported. initially 0
- *pendingBlocks*: stores for each height which peer was
  queried. initially nil for each height
- *receivedBlocks*: stores for each height which peer returned
  it. initially nil
- *blockstore*: stores for each height greater than
    *startBlock.Height*, the block of that height. initially nil for
    all heights

> - *peerTimeStamp*: stores for each peer the last time a block was received

Reviewer: There is also pendingTime variable in the code where time of the block request is stored. It is used as part of tryPrune function but only for the peer that is responsible for current height.

Author: I added at this point

- *pendingTime*: stores for each peer the last time a block was requested

and later, I updated:

#### **[FS-V2-PEER-REMOVE]**:
Periodically, *peerTimeStamp* and *peerRate* and *pendingTime* are
analyzed. 
If a peer *p*
has not provided a block recently (check of *peerTimeStamp[p]*) or it
has not provided sufficiently many data (check of *peerRate[p]*), then
*p* is removed from *peerIDs*. In addition, *pendingTime* is used to
estimate whether the peer that is responsible for current height is
still connected.

> the verification, but can also guide the implementation.

#### **[FS-VAR-STATE-INV]**:
It is always the case that *state* corresponds to the application state of the
blockchain of that height, that is, *state = chain[height -
1].AppState*; *chain* is defined in
[**[TMBC-SEQ]**][TMBC-SEQ-link].

#### **[FS-VAR-PEER-INV]**:
It is always the case that the set *peerIDs* only contains nodes that
have not yet misbehaved (by sending wrong data or timing out).

#### **[FS-VAR-BLOCK-INV]**:
For *startBlock.Height <= i < height - 1*, let *b(i)* be the block with
height *i* in *blockstore*, it always holds that
*VerifyCommit(b(i), b(i+1).Commit) = true*. This means that *height*
can only be incremented if all blocks with lower height have been verified.

## FastSync V2

<!--
> Basic data structures. Simplified, so that we can focus on the distributed
algorithm here. If existing: link to Tendermint data structures, and mentioned
if details were omitted. 
---->

### Outline

<!--
> Describe solution (in English), decomposition into functions, where communication to other components happens.
---->

The protocol is described in terms of functions that are triggered by

> (external) events. The implementation uses a scheduler and a

Reviewer: I am not sure that this sentence is correct. Maybe we just need to explain that some messages are sent periodically (statusRequest) and some actions that lead to sending messages are also triggered periodically. We just need to explain basically execution model.

Author: At some point we said we want to keep this underspecified on purpose, now it seems that this and the other comments target at specifying V2 more precisely. We should fix what level of abstraction we are aiming for.

de-multiplexer to deal with communicating with peers and to
trigger the execution of these functions:

- `QueryStatus()`: regularly (currently every 10sec; necessarily
  interval greater than *2 Delta*) queries all peers from *peerIDs*
  for their current height [TMBC-CORR-FULL]. It does so
  by calling `Status(n)` remotely on all peers *n*.
  
- `CreateRequest`: regularly checks whether certain blocks have no
  open request. If a block does not have an open request, it requests
  one from a peer. It does so by calling `Block(n,h)` remotely on one
  peer *n* for a missing height *h*.
  

> > We have left the strategy how peers are selected unspecified, and

Reviewer: Maybe we should explain the strategy used as it is part of the solution. It basically selects a peer with the minimum number of pending requests that can serve given height h (peerHeight >= h).

Author: I added the clarification for V2 to the comment.

> the currently existing different implementations of Fastsync differ
> in this aspect. 

The functions `Status` and `Block` are called by asynchronous
RPC. When they return, the following functions are called:

- `OnStatusResponse(addr Address, height int64)`: The full node with
  address *addr* returns its current height. The function updates the height
  information about *addr*, and may also increase *TargetHeight*.
  
- `OnBlockResponse(addr Address, b Block)`. The full node with
  address *addr* returns a block. It is added to *blockstore*. Then
  the auxiliary function `Execute` is called.

- `Execute()`: Iterates over the *blockstore*.  Checks soundness of
  the blocks, and
  executes the transactions of a sound block and updates *state*. 

> In addition to the functions above, the following two features are
> implemented in Fastsync V2

#### **[FS-V2-PEER-REMOVE]**:
Periodically, *peerTimeStamp* and *peerRate* are analyzed. If a peer *p*
has not provided a block recently (check of *peerTimeStamp[p]*) or it
has not provided sufficiently many data (check of *peerRate[p]*), then
*p* is removed from *peerIDs*.
  
#### **[FS-V2-TIMEOUT]**:

> *Fastsync V2* starts a timeout whenever a block is

Reviewer: To be precise, implementation is slightly different. Node remember the last point in time when a block has been executed. Then there is an action that is triggered periodically and if no progress has been made for at least syncTimeout, i.e., (now - lastUpdate > syncTimeout) then fastSync terminates.

Author: We can describe it more precisely, but I would consider this an implementation detail of how it is checked whether a timeout occurred.

executed (that is, when the height is incremented). If the timeout expires
before the next block is executed, *Fastsync* terminates.

> We say that if *peerIDs* is empty upon termination, then *Fastsync* terminates

Reviewer: Not sure about this. Note that you can terminate in this case with a peerset not empty, and I am not sure if it is still considered success. My understanding was that success is in case we sync up to maxHeight and failure is when we terminate because of this timeout.

Author: I thought that in the synchronous system your scenario is not possible. But I guess I didn't make explicit that the timeout must be sufficiently large.

with failure, otherwise it terminates successfully.

### Details

<!--
> Function signatures followed by pseudocode (optional) and a list of features (required):
> - Implementation remarks (optional)
>   - e.g. (local/remote) function called in the body of this function
> - Expected precondition
> - Expected postcondition
> - Error condition
---->

```go
func QueryStatus()
```
- Expected precondition
    - peerIDs initialized and non-empty
- Expected postcondition
    - call asynchronously `Status(n)` at each peer *n* in *peerIDs*.
- Error condition
    - fails if precondition is violated
----

```go
func OnStatusResponse(addr Address, ht int64)
```
- Comment
    - *ht* is a height
	- peers can provide the status without being called
- Expected precondition
    - *peerHeights(addr) <= ht*
- Expected postcondition
    - *peerHeights(addr) = ht*
	- *TargetHeight* is updated
- Error condition
    - if precondition is violated: *addr* not in *peerIDs* (that is,
      *addr* is removed from *peerIDs*)
- Timeout condition

>     - if `OnStatusResponse(addr, ht)` was not invoked within *2 Delta* after

Reviewer: Note that in V2 this is not the case, i.e., it is not tracked when statusResponse arrived. Therefore, peers are not removed upon late statusReply.

	`Status(addr)` was called:  *addr* not in *peerIDs*
----

```go
func CreateRequest
```
- Expected precondition
    - *height < TargetHeight*
	- *peerIDs* nonempty
- Expected postcondition
    - Function `Block` is called remotely at a peer *addr* in peerIDs 
	  for a missing height *h*  

> 	  *Remark:* different implementations may have different

Reviewer: As this spec is about v2, we should probably details strategy implemented in v2.

      strategies to balance the load over the peers

>     - *pendingblocks(h) = addr*

Reviewer: It also remember time at which request is made, but don't know if we want to enter that level of details as you capture it at more high level above.

----

```go
func OnBlockResponse(addr Address, b Block)
```
- Comment
    - if after adding block *b*, blocks of heights *height* and

>       *height + 1* are in *blockstore*, then `Execute` is called

Reviewer: Execute is in v2 triggered periodically, i.e., it is not triggered by block reception.

- Expected precondition
    - *pendingblocks(b.Height) = addr*

> 	- *b* satisfies basic soundness  

Reviewer: blockstore(b.Height) = nil (empty)

- Expected postcondition
    - if function `Execute` has been executed without error or was not
      executed:
        - *receivedBlocks(b.Height) = addr*
        - *blockstore(b.Height) = b*

>         - *peerTimeStamp[addr]* is set to a time between invocation and

Reviewer: peerTimestamp[addr] is set to the current time in v2, but maybe you want to capture here more high level concept.

          return of the function.
        - *peerRate[addr]* is updated according to size of received

>           block

Reviewer: ```suggestion
          block and time it has passed between current time and last block received from this peer (addr)
```

- Error condition

>     - if precondition is violated: *addr* not in *peerIDs*; reset

Reviewer: Removing a peer triggers cleaning data structure such that all pending/received but not processed requests/blocks are removed.

	*pendingblocks(b.Height)* to nil;
- Timeout condition

>     - if `OnBlockResponse(addr, b)` was not invoked within *2 Delta* after

Reviewer: In v2, there is no such check. Removal happens periodically but only for the current height this kind of check is implemented. For other peers, it looks at peer rate and difference between current time and last block reception.

	`Block(addr,h)` was called for *b.Height = h*: *addr* not in *peerIDs*
----

```go
func Execute()
```
- Comments
    - none

> - Expected precondition

Reviewer: Precondition can be that blocks for height and height+1 are in the blockStore, and that VerifyCommit (a,b.Commit) = true. Postcondition is that block at height is executed, state updated and height incremented. There are two error cases: 1) blocks are not present, do nothing, or 2) blocks are present but VerifyCommit (a,b.Commit) = false. Then both blocks are removed and the corresponding peers are removed.

	- application state is the one of the blockchain at height
      *height - 1*
	- **[FS-V2-Verif]** for any two blocks *a* and *b* from
	*receivedBlocks*: if
	  *a.Height + 1 = b.Height* then *VerifyCommit (a,b.Commit) = true*
- Expected postcondition
    - Any two blocks *a* and *b* violating [FS-V2-Verif]:
	  *a* and *b* not in *blockstore*; nodes with Address 
	  receivedBlocks(a.Height) and receivedBlocks(b.Height) not in peerIDs
	- height is updated height of complete prefix that matches the blockchain
	- state is the one of the blockchain at height *height - 1*
- Error condition
    - none
----

## Analysis of Fastsync V2

####  **[FS-ISSUE-KILL]**:
If two blocks are not matching [FS-V2-Verif], `Execute` dismisses both
blocks and removes the peers that provided these blocks from
*peerIDs*. If block *a* was correct and provided by a correct peer *p*,
and block b was faulty and provided by a faulty peer, the protocol
- removes the correct peer *p*, although it might be useful to

>   download blocks from it in the future

Reviewer: Maybe it's better to link it with failed invariant (we don't download all blocks from correct peers).

	   
4. Let *f(t)* be the number of faulty peers in *peerIDs* at time *t*;  
   *f(begin) = fmax*.
	   
5. Let t_i be the sequence of times `OnBlockResponse(addr,b)` is
   invoked or times out with *b.Height = height + 1*.
   
6. By 3., 
   - (a). *t_1 <= begin + 2 Delta* 
   - (b). *t_{i+1} <= t_i + 2 Delta* 

7. By an inductive argument we prove for *i > 0* that

   - (a). *height(t_{i+1}) > height(t_i)*, or
   - (b). *f(t_{i+1}) < f(t_i))* and *height(t_{i+1}) = height(t_i)*  

   Argument: if the peer is faulty and does not return a block, the
   peer is removed, if it is faulty and returns a faulty block
   `SequentialVerify` removes the peer (b). If the returned block is OK,
   height is increased (a).
	 
8. By 2. and 7., faulty peers can delay incrementing the height at
   most *fmax* times, where each time "costs" *2 Delta* seconds. We
   have additional *2 Delta* initial offset (3a) plus *2 Delta* to get
   all missing blocks after the last fault showed itself. (This
   assumes that an arbitrary number of blocks can be obtained and
   checked within one round-trip 2 Delta; which either needs
   conservative estimation of Delta, or a more refined analysis). Thus

>    we reach the *targetHeight* and terminate by time *end*.

Reviewer: I like this analysis. This was my intuition, but I wasn't able to formalise it. I am just wondering if faulty peers sending statusResponse for nonexistent blocks can prevent us from termination. I am wondering also if the fact that correct peers are not fully up to date with the blockchain matters. It seems that what this proposal ensures is that once we hit max height of correct peers, if faulty peers advertise higher height they will eventually be removed as *2 Delta < ETIME*.

      the protocol variables that the implementation should maintain.
	  
    - [FastSync V2](#FastSync-V2): gives an outline of the solution,
      and details of the functions used (with preconditions,
      postconditions, error conditions).
	  
    - [Analysis of Fastsync V2](#Analysis-of-Fastsync-V2): describes
      undesireable scenarios of Fastsync V2, and why they violate
      desireable temporal logic specification in an unreliable
      distributed system.

- [Part III](#part-iii---suggestions-for-an-improved-fastsync-implementation):
  some suggestions on how to address the issues in the future.
  
   - [Desirable Temporal Properties](#Desirable-Temporal-Properties):
     we give temporal properties that describe a fault-tolerant
     version
	 of FastSync.
	 
   - [Suggestions](#Suggestions)  to address the issues discussed in the analysis.
  

In this document we quite extensively use tags in order to be able to
reference assumptions, invariants, etc. in future communication. In
these tags we frequently use the following short forms:

- TMBC: Tendermint blockchain
- SEQ: for sequential specifications
- FS: Fastsync
- LIVE: liveness
- INV: invariant

> - TERM: termination

Reviewer: What is difference between liveness and termination? I would expect that it is the same and that one abbreviation is sufficient. On the other side we don't have abbreviation for safety properties.

Author: As you noted below, TERM was mostly addressing safety properties upon termination. I will remove TERM here, as we have changed the tags

perhaps within 10 blocks away from the current height of the blockchain.
Fastsync should bring a node within this range.

### Blockchain

> We will briefly list some of the notions
> of blockchains as required for this specification. More details can
> be found  [here][block].

#### **[TMBC-HEADER]**:
A set of blockchain transactions is stored in a data structure called
*block*, which contains a field called *header*. (The data structure
*block* is defined [here][block]).  As the header contains hashes to
the relevant fields of the block, for the purpose of this
specification, we will assume that the blockchain is a list of
headers, rather than a list of blocks. 

#### **[TMBC-SEQ]**:

The Tendermint blockchain is a list *chain* of headers. 

#### **[TMBC-SEQ-GROW]**: 

During operation, new headers may be appended to the list one by one.

> In the following, *ETIME* is a lower bound
> on the time interval between the times at which two
> successor blocks are added. 

#### **[TMBC-SEQ-APPEND-E]**: 
If a header is appended at time *t* then no additional header will be
appended before time *t + ETIME*.

#### **[TMBC-AUTH-BYZ]**:

> The authenticated Byzantine model assumes that no node (faulty or

Reviewer: ```suggestion
We assume the authenticated Byzantine model in which no node (faulty or
```

correct) may break digital signatures, but otherwise, no additional

> assumption is made about the internal behavior of faulty full

Reviewer: Full node is not defined at this point in time. Maybe it's sufficient to say here just faulty nodes.

nodes. That is, faulty nodes are only limited in that they cannot forge
messages.

<!-- The authenticated Byzantine model assumes [TMBC-Sign-NoForge] and -->
<!-- [TMBC-FaultyFull], that is, faulty nodes are limited in that they -->
<!-- cannot forge messages [TMBC-Sign-NoForge]. -->

> > We observed that in the existing documentation the term

Reviewer: What documentation? Tendermint documentation?

Author: yes.

> *validator* refers to both a data structure and a full node that
> participates in the distributed computation. Therefore, we introduce
> the notions *validator pair* and *validator node*, respectively, to
> distinguish these notions in the cases where they are not clear from
> the context.

#### **[TMBC-VALIDATOR-PAIR]**:

Given a full node, a 
*validator pair* is a pair *(address, voting_power)*, where 
  - *address* is the address (public key) of a full node, 
  - *voting_power* is an integer (representing the full node's

>   voting power in a certain consensus instance).

Reviewer: ```suggestion
  voting power in a given consensus instance).
```

  
> In the Golang implementation the data type for *validator

> > pair* is called `Validator`

Reviewer: ```suggestion
> pair* is called `Validator`.
```

#### **[TMBC-VALIDATOR-SET]**:

A *validator set* is a set of validator pairs. For a validator set
*vs*, we write *TotalVotingPower(vs)* for the sum of the voting powers
of its validator pairs.

#### **[TMBC-CORRECT]**:
We define a predicate *correctUntil(n, t)*, where *n* is a node and *t* is a 
time point. 
The predicate *correctUntil(n, t)* is true if and only if the node *n* 
follows all the protocols (at least) until time *t*.

#### **[TMBC-TIME-PARAMS]**:
A blockchain has the following configuration parameters:

>  - *unbondingPeriod*: a time duration.

Reviewer: It seems that we don't use unbonding period term.

Author: I think we should still use it as people know the term?

Reviewer: Ok, fair enough.

 - *trustingPeriod*: a time duration smaller than *unbondingPeriod*.

#### **[TMBC-FM-2THIRDS]**:
If a block *h* is in the chain,
then there exists a subset *CorrV*
of *h.NextValidators*, such that:
  - *TotalVotingPower(CorrV) > 2/3
    TotalVotingPower(h.NextValidators)*; 
  - For every validator pair *(n,p)* in *CorrV*, it holds *correctUntil(n,
    h.Time + trustingPeriod)*.

#### **[TMBC-CORR-FULL]**: 
Every correct full node locally stores a prefix of the
current list of headers from [**[TMBC-SEQ]**][TMBC-SEQ-link].

> ## Informal Problem statement

Reviewer: ```suggestion
## Fastsync Informal Problem statement
```

A full node has as input a block of the blockchain at height *h* and
the corresponding application state (or the prefix of the current
blockchain until height *h*). It has access to a set *peerIDs* of full

> nodes called *peers* that it knows of.  The full node uses the peers

Reviewer: Full node is never defined if I am not wrong.

to read blocks of the Tendermint blockchain (in a safe way, that is,
it checks the soundness conditions), until it has read the most recent
block and then terminates.

## Sequential Problem statement

*Fastsync* gets as input a block of height *h* and the corresponding
application state *s* that corresponds to the block and state of that
height of the blockchain, and produces
as output (i) a list *L* of blocks starting at height *h* to some height
*terminationHeight*, and (ii) the application state when applying the
transactions of the list *L* to *s*. Fastsync has to satisfy the following

> properties [FS-SEQ-?]:

Reviewer: ```suggestion
properties:
```

#### **[FS-SEQ-LIVE]**: 
*Fastsync* eventually terminates.

 

> #### **[FS-SEQ-TERM]**:

Reviewer: This is safety property, so the tag could probably be FS-SAFE-START.

Let *bh* be the height of the blockchain at the time *Fastsync*
starts. By assumption we have *bh >= h*.
When *Fastsync* terminates, it outputs a list of all blocks from

> height *h* to some height *terminationHeight >= bh - 1*.

Reviewer: I don't get this property. Not sure why you need to use both bh and h as starting height.

> The above property is independent of how many blocks are added to the
> blockchain while Fastsync is running. It links the target height to the
> initial state. If Fastsync has to catch-up many blocks, it would be
> better to link the target height to a time close to the

> > termination. This is capture by the following specification:

Reviewer: ```suggestion
> termination. This is captured by the following specification:
```

> #### **[FS-SEQ-TERM-SYNC]**:

Reviewer: FS-SAFE-SYNC?

Let *eh* be the height of the blockchain at the time *Fastsync*
terminates.  There is a constant *D >= 1* such that when *Fastsync*
terminates, it outputs a list of all blocks from height *h* to some
height *terminationHeight >= eh - D*.

#### **[FS-SEQ-INV]**:
Upon termination, the application state is the one that corresponds to
the blockchain at height *terminationHeight*.

> As the block of height *h + 1* is needed to verify the block of

> > height *h* we highlight th following clarification of the

Reviewer: ```suggestion
> height *h* we highlight the following clarification of the
```

> termination height.

#### **[FS-SEQ-HEIGHT]**: 
The returned value *terminationHeight* is the height of the block with the largest
height that could be verified. In order to do so, *Fastsync* needs the

> Commit of the block at height  *terminationHeight + 1* in the blockchain.

Reviewer: ```suggestion the block at height  *terminationHeight + 1* in the blockchain.
```

> Commit of the block at height  *terminationHeight + 1* in the blockchain.

Reviewer: ```suggestion
the block at height  *terminationHeight + 1* in the blockchain.
```

# Part II - Protocol view

## Environment/Assumptions/Incentives

#### **[FS-A-NODE]**:
We consider a node *FS* that performs *Fastsync*.

#### **[FS-A-PEER-IDS]**:

> *FS* has access to a set *peerIDs* of IDs (public keys) of peers (full

Reviewer: ```suggestion
*FS* has access to a set *peerIDs* of IDs (public keys) of peers
```

>      nodes). During the execution of *Fastsync*, another protocol (outside

Reviewer: ```suggestion
     . During the execution of *Fastsync*, another protocol (outside
```

     of this specification) may add new IDs to *peerIDs*.

#### **[FS-A-PEER]**:
Peers can be faulty, and we do not make any assumptions about the number or
ratio of correct/faulty nodes. Faulty processes may be Byzantine
according to [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link].

#### **[FS-A-VAL]**:

> The system satisfies [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link] and [**[TMBC-FM-2THIRDS]**][TMBC-FM-2THIRDS-link]. Thus, there is a

Reviewer: This is not really clear in my view. We have a node and peers, and I guess that with this property we want to make a link to validators and Tendermint blockchain. I am not sure what is the easiest way to introduce this without going into details of explaining validators and difference between validators and full nodes. From the fast sync view point the whole blockchain business can be seen as application specific stuff, and it is only relevant as part of block validation in the proposed solution.

Author: If we want to say what blocks to accept and execute, we have to state that there is a reference block chain that has these blocks. It is not clear how to specify reasonable invariants without having that reference. It is the validity property of Fastsync.   
Later, in our suggestions to use the lightclient-like verification (at the end of the document) we need this as it is based on the tendermint failure model.

bounded in time: there is a message end-to-end delay *Delta* such that
if a message is sent at time *t* by a correct process to a correct
process, then it will be received and processed by time *t +
Delta*. This implies that we need a timeout of at least *2 Delta* for
remote procedure calls to ensure that the response of a correct peer
arrives before the timeout expires.

## Distributed Problem Statement

### Design choices

#### Two Kinds of Termination

We do not assume that there is a correct full node in
*peerIDs*. Under this assumption no protocol can guarantee the combination
of the properties [FS-SEQ-LIVE] and
[FS-SEQ-TERM] and [FS-SEQ-TERM-SYNC] described in the sequential
specification above. Thus, in the (unreliable) distributed setting, we
consider two kinds of termination (successful and failure) and we will
specify below under what (favorable) conditions *Fastsync* ensures to
terminate successfully, and satisfy the requirements of the sequential
problem statement:

#### **[FS-DISTR-TERM]**:
*Fastsync* eventually terminates: it either *terminates successfully* or it  *terminates with failure*.

#### Remote Functions

> The Tendermint Full Node exposes the following functions over

Reviewer: We haven't defined what is a Tendermint full node. Can we just say that peers expose the following functions over RPC?

   - RPC to full node *addr*
- Expected precondition
  - 'height` is less than or equal to height of the peer
- Expected postcondition
  - if *addr* is correct: Returns the block of height `height`
  from the blockchain. [FS-A-COMM]
  - if *addr* is faulty: Returns arbitrary block [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
- Error condition
  - if *addr* is correct: precondition violated. [FS-A-COMM]
  - if *addr* is faulty: arbitrary error (including timeout). [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
----

### Temporal Properties

#### Fairness

As mentioned above, without assumptions on the correctness of some
peers, no protocol can achieve the required specifications. Therefore, 
we sometimes consider the following (fairness) constraint in the
safety and liveness properties below:

#### **[FS-SOME-CORR-PEER]**:
Initially, the set *peerIDs* contains at least one correct full node.

> While in principle the above condition can be part of a sufficient
> condition to solve [FS-SEQ-LIVE] and
> [FS-SEQ-TERM] and [FS-SEQ-TERM-SYNC], we will see below that the
> current implementation of Fastsync (V2) requires the following (much
> stronger) requirement

#### **[FS-ALL-CORR-PEER]**:

> At all times, the set *peerIDs* contains only correct full nodes.

Reviewer: We need to check if some properties of v2 hold in the crash tolerant model.

#### Safety

<!--
> safety specifications / invariants in English 
---->

#### **[FS-VC-ALL-CORR-NONABORT]**:
Under [FS-ALL-CORR-PEER], *Fastsync* never terminates with failure.

#### **[FS-VC-STATE-INV]**:

> If *FastSync* terminates successfully at height *terminationHeight*, then the

Reviewer: Isn't this always the case? Difference between success and failure in my view is only with respect to the quality of the result, i.e, about number of synced blocks.

Author: I wanted to recall *terminationheight* here, but we can do without, I removed it.

application state is the one that corresponds to the blockchain at
height *terminationHeight*.

#### **[FS-VC-BLOCKS-INV]**:

> If *FastSync* terminates successfully at height *terminationHeight*, then the

Reviewer: The same comment as above.

Author: removed it

returned list of blocks  is the one that corresponds to the blocks of
the
blockchain.

> As this specification does
> not assume that a correct peer is at the most recent height
> of the blockchain (it might lag behind), the property [FS-SEQ-TERM]
> cannot be ensured in an unreliable distributed setting. We consider
> the following relaxation. (Which is typically sufficient for
> Tendermint, as the consensus reactor then synchronizes from that
> height.)

#### **[FS-VC-CORR-INV]**:
Under [FS-ALL-CORR-PEER], let *maxh* be the maximum 
height of a correct peer [**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link]
in *peerIDs* at the time *Fastsync* starts. If *FastSync* terminates

> successfully, it is at some height *terminationHeight >= maxh - 1*.

Reviewer: I might be wrong but I haven't seen definition of termination height. Is that the height of the last verified and executed block?

Author: you are wrong ;-) we have the following:
#### **[FS-SEQ-HEIGHT]**: 
The returned value *terminationHeight* is the height of the block with the largest
height that could be verified. In order to do so, *Fastsync* needs the
block at height  *terminationHeight + 1* of the blockchain.

> > The above property is independent of how many blocks are added to the

Reviewer: It is also independent of addition of new peers. As peers can be removed by external component (not just added), we probably need additional assumption that a correct peer is not removed from the peerIds by the external component.

> blockchain (and learned by the peers) while *Fastsync* is running. It
> links the target height to the initial state. If *Fastsync* has to
> catch-up many blocks, it would be better to  link the
> target height to a time close to the termination. This is captured by
> the following specification:

#### **[FS-VC-CORR-INV-SYNC]**:
Under [FS-ALL-CORR-PEER], there exists a constant time interval *TD*, such
that if *term* is the time *Fastsync* terminates and
*maxh* be the maximum height of a correct peer
[**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link] in *peerIDs* at the time

> *term - TD*, then if *FastSync* terminates successfully, it is at

Reviewer: Sequential property (FS-SEQ-TERM-SYNC) refers to a logical diff (parameter D that is expressed as lag in number of heights), while here we introduce something similar but not the same with time parameter TD. Can we somehow align this better or it is fine as sequential spec is defined without reference to system parameters.

Author: I agree the issue here is not explicitly discussed. The problem is that in the case the peers are themselves synchronized to the blockchain (they are at the current height), then *TD* together with LTIME in [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]  can be used to bound $D$. If the peers themselves lag behind and catch-up during the *TD* time, we cannon bound anything. Perhaps it is enough to add an explanation at this point to make it clear. What do you think?
